source/layers/validation/checkers/events_deadlock/zel_events_deadlock_checker.cpp

```diff
@@ -0,0 +1,256 @@
+    ze_event_handle_t *phWaitEvents        ///< [in][optional][range(0, numWaitEvents)] handle of the events to wait
+                                           ///< on before launching
+) {
+
+    std::cout << "eventsDeadlockChecker: zeCommandListAppendMemoryCopyPrologue" << std::endl;
+    printf("\t\tNumber of wait events: %d\n", numWaitEvents);
+    for (uint32_t i = 0; i < numWaitEvents; i++) {
+        printf("\t\tphWaitEvents[%d] pointer address: %p\n", i, phWaitEvents[i]);
+    }
+
+    int this_action_new_node_id = invalidDagID;
+
+    if (hSignalEvent != nullptr) {
+
+        if (eventToDagID.find(hSignalEvent) != eventToDagID.end()) {
+            if (eventToDagID[hSignalEvent] != invalidDagID) {
+                // This event already exists in the DAG. Get the DAG node ID.
+                // For example when there is indeed a deadlock it would have already been created.
+                this_action_new_node_id = eventToDagID[hSignalEvent];
+                std::cout << "\tFound event in eventToDagID: hSignalEvent = " << hSignalEvent << ", this_action_new_node_id = " << this_action_new_node_id << std::endl;
+            }
+        }
+    }
+
+    if (this_action_new_node_id == invalidDagID) {
+        // Create node in DAG
+        this_action_new_node_id = nextDagID;
+        nextDagID++;
+
+        // Now we know where the hSignalEvent points from/out in the DAG. Update the eventtoDagID map.
+        eventToDagID[hSignalEvent] = this_action_new_node_id;
+        std::cout << "\tUpdated eventToDagID: hSignalEvent = " << hSignalEvent << ", this_action_new_node_id = " << this_action_new_node_id << std::endl;
+    }
+
+    // Add this action to the actionToDagID map
+    std::string action = "MemoryCopy";
+    dagIDToAction[this_action_new_node_id] = actionAndSignalEvent(action, hSignalEvent);
+
+    // Form the dependency in the DAG
+    if (numWaitEvents != 0) {
+
+        for (uint32_t i = 0; i < numWaitEvents; i++) {
+            if (eventToDagID.find(phWaitEvents[i]) != eventToDagID.end()) {
+
+                int dagID = eventToDagID[phWaitEvents[i]];
+                if (dagID == invalidDagID) {
+                    // Create a new node in the DAG for this wait event. That action will be created some time in the future.
+                    int new_future_node_id = nextDagID;
+                    nextDagID++;
+                    eventToDagID[phWaitEvents[i]] = new_future_node_id;
+
+                    dagID = new_future_node_id;
+                }
+
+                // Add edge (dependency) from dagID to this_action_new_node_id in the DAG.
+                std::cout << "\tDAG: Adding edge from " << dagID << " to " << this_action_new_node_id << std::endl;
+            } else {
+                std::cerr << "eventsDeadlockChecker: zeCommandListAppendMemoryCopyPrologue: Error: Wait event not found in eventToDagID map" << std::endl;
+                std::terminate();
```

Comment:
If something goes wrong we sohuld return an error instead of terminating.

We also need to make sure any error messages we log are understandable to someone that is not familiar with the internals here